test/local/routes/devices-and-sessions.js

```diff
@@ -498,6 +509,155 @@ describe('/account/devices/notify', function () {
+      getOAuthToken: sinon.spy(() => {
+        return Promise.resolve({access_token: 'tok'})
+      })
+    })
+    const mockPushbox = mocks.mockPushbox({
+      retrieve() {
+        const error = new Error()
+        error.message = 'Boom!'
+        error.statusCode = 500
+        return Promise.reject(error)
+      }
+    })
+    mockRequest.query = {
+      index: 2
+    }
+    const route = getRoute(makeRoutes({
+      customs: mockCustoms,
+      log: mockLog,
+      oAuthMint: mockOAuthMint,
+      pushbox: mockPushbox
+    }), '/account/device/messages')
+
+    return runTest(route, mockRequest).then(() => {
+      assert.ok(false, 'should not go here')
+    }, (err) => {
+      assert.equal(err.message, 'Boom!')
+      assert.equal(err.statusCode, 500)
+    })
+  })
+})
+
+describe('/account/devices/messages', function () {
+  const uid = uuid.v4('binary').toString('hex')
+  const mockLog = mocks.mockLog()
+  const mockRequest = mocks.mockRequest({
+    log: mockLog,
+    devices: [
+      {
+        id: 'bogusid1',
+        type: 'mobile'
+      },
+      {
+        id: 'bogusid2',
+        type: 'desktop',
+      }
+    ],
+    credentials: {
+      uid: uid
+    }
+  })
+  const mockPush = mocks.mockPush()
+  const mockCustoms = mocks.mockCustoms()
+
+  it('store messages using the pushbox service', () => {
+    const pushboxToken = '558f9980ad5a9c279beb52123653967342f702e84d3ab34c7f80427a6a37e2c0'
+    const mockOAuthMint = mocks.mockOAuthMint({
+      getOAuthToken: sinon.spy(() => {
+        return Promise.resolve({access_token: pushboxToken})
+      })
+    })
+    const mockPushbox = mocks.mockPushbox({
+      store: sinon.spy(() => {
+        return Promise.resolve('https://foo.bar')
+      })
+    })
+    const to = 'bogusid1'
+    const topic = 'sendtab'
+    const data = 'eyJmb28iOiAiYmFyIn0'
+    mockRequest.payload = {
+      to,
+      topic,
+      data
+    }
+    const route = getRoute(makeRoutes({
+      customs: mockCustoms,
+      log: mockLog,
+      oAuthMint: mockOAuthMint,
+      push: mockPush,
+      pushbox: mockPushbox
+    }), '/account/devices/messages')
+
+    return runTest(route, mockRequest).then(() => {
+      assert.equal(mockPushbox.store.callCount, 1, 'pushbox was called')
+      let args = mockPushbox.store.args[0]
+      assert.equal(args.length, 5)
+      assert.equal(args[0], uid)
+      assert.equal(args[1], to)
+      assert.equal(args[2], topic)
+      assert.equal(args[3], data)
+      assert.equal(args[4], pushboxToken)
+
+      assert.equal(mockPush.notifyMessageReceived.callCount, 1,
+        'notifyMessageReceived was called')
+      args = mockPush.notifyMessageReceived.args[0]
+      assert.equal(args.length, 5)
+      assert.equal(args[0], uid)
+      assert.deepEqual(args[1], {id: 'bogusid1', type: 'mobile'})
+      assert.equal(args[2], 'https://foo.bar')
+      assert.equal(args[3], topic)
+      assert.equal(args[4], 'sendTab')
+    })
+  })
+})
```

Comment:
Is it worth worth adding at least one remote test in `test/remote/device_tests.js` to exercise this code end-to-end, even if it's just the happy case of "post a message, read it back"?

Of course, that does raise the question of whether we need to mock out pushbox along the lines of `./test/oauth_helper.js` and friends....which may not be worth the effort in practice.  I'll gladly defer that to your best judgement, but wanted to mention it here.

I also wonder what that means for fxa-local-dev, and whether we want to run a mock pushbox server there fore completeness...@vladikoff thoughts?  Maybe we can pull it in via docker image?

Reply:
I think we should add a pushbox_helper.js once the exact specs of the rustbox server are stabilized (for edge cases like unknown deviceId, unknown index etc)